### PR TITLE
oop-special_class_method-hard-q14

### PR DESCRIPTION
**Pull request by KaiKhademi**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/special_class_method/hard/q14/QuestionText.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+As part of a larger virtual reality program, you want to build more life-like objects. Make a Java class called 
+**Bottle** that has an attribute called **isFull** which is initialized to **false** in the constructor.
+
+### Bottle Class:
+
+Your `Bottle` class must have the following items:
+
+- Attributes:
+    - A `boolean` called **isFull**
+    - These are not visible from other classes
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/QuestionText.md:14]: make this singular not plural

File: src/oop/special_class_method/hard/q14/QuestionText.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    public Bottle()
+    isFull = true;
+    isFull = FALSE;
+    isFull = TRUE;
+    {
+    }
+    private Bottle
+    private String toString( String str )
+    private boolean isFull;
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/QuestionText.md:34]: this line is already used in the solution

File: src/oop/special_class_method/hard/q14/QuestionText.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    - A `boolean` called **isFull**
+    - These are not visible from other classes
+
+- Constructor:
+    - a `Bottle` constructor that initialize the **isFull** attribute to false.
+
+- Methods:
+    - a `setIsFull` method which accepts a `boolean` and uses it to set the value of the **isFull** attribute
+    - a `getIsFull` method which returns the value stored in the **isFull** attribute
+    - These methods must be visible from other classes.
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/QuestionText.md:22]: there should be a toString method described here

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    private ObjectTest testClass;
+
+    @BeforeEach
+    public void setUp() throws Throwable {
+        String classString = "oop.special_class_method.hard.q14." + className;
+        this.testClass = new ObjectTest(classString);
+    }
+
+    @Test
+    public void alienClassHasCorrectAttributes() {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/MainTest.java:31]:
```suggestion
    public void bottleClassHasCorrectAttributes() {
```

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        assertTrue(testClass.hasField(attributeName),
+                "Your " + className + " class is missing the " + attributeName + " attributes, or it is spelled incorrectly.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/MainTest.java:33]:
```suggestion
                "Your " + className + " class is missing the " + attributeName + " attribute, or it is spelled incorrectly.");
```

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName, testInstance);
+        assertTrue(testClass.hasModifier(getAttributeMethodName, null, "public"),
+                "Your " + getAttributeMethodName + " method does not have the correct visibility modifier.");
+        _assertEquals(value, getMethodOutput,
+                "Your " + getAttributeMethodName + " method does not return the value of the " + attributeName + " attribute.");
+    }
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctSetYearMadeMethod(boolean initialiValue, boolean updatedValue) throws Throwable {
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/MainTest.java:81]:
```suggestion
    public void correctSetIsFullMethod(boolean initialiValue, boolean updatedValue) throws Throwable {
```

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        testClass.setFieldValue(testInstance, initialValue, attributeName);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName, testInstance);
+        _assertEquals(initialValue, getMethodOutput,
+                "Your " + getAttributeMethodName + " method does not return the value of the " + attributeName + " attribute.");
+        Object[][] setMethodArguments = {
+                {updatedValue, boolean.class}
+        };
+        testClass.callMethod(setAttributeMethodName, setMethodArguments, testInstance);
+        _assertEquals(updatedValue, testClass.callMethod(getAttributeMethodName, testInstance),
+                "Your " + getAttributeMethodName + " method does not correctly update the value of " + setAttributeMethodName + " after using the " + getAttributeMethodName + " method.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/MainTest.java:129]:
```suggestion
                "Your " + getAttributeMethodName + " method does not correctly return the value of " + attributeName + " after using the " + setAttributeMethodName + " method.");
```

**Review by SamamthaH — CHANGES_REQUESTED** [review]

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                "Your " + className + " constructor does not initialize the " + attributeName + " attribute to the correct value.");
+    }
+
+
+    @ParameterizedTest
+    @MethodSource("inputProvider")
+    public void correctGetIsFullMethod(boolean value) throws Throwable {
+        Object testInstance = testClass.createInstance();
+        testClass.setFieldValue(testInstance, value, attributeName);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName, testInstance);
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/MainTest.java:72]: use .hasMethod first

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+        Object testInstance = testClass.createInstance();
+        testClass.setFieldValue(testInstance, initialiValue, attributeName);
+        Object[][] setMethodArguments = {
+                {updatedValue, boolean.class}
+        };
+        Class<?>[] methodModifierClasses = {
+                boolean.class
+        };
+        assertTrue(testClass.hasMethod(setAttributeMethodName, methodModifierClasses, Void.TYPE),
+                "Your " + setAttributeMethodName + " method does not have the correct name, return type, or parameters.");
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/MainTest.java:91]: return type should be split into its own assert

File: src/oop/special_class_method/hard/q14/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
+# oop-special_class_method-hard-q14
+
+## Question Text
+
+As part of a larger virtual reality program, you want to build more life-like objects. Make a Java class called 
+**Bottle** that has an attribute called **isFull** which is initialized to **false** in the constructor.
```
Comment (SamamthaH) [inline src/oop/special_class_method/hard/q14/QuestionText.md:6]: get rid of duplicate class information

**Review by SamamthaH — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]
See feedback on other questions in this category and apply it to this PR.

File: src/oop/special_class_method/hard/q14/QuestionText.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
+# oop-special_class_method-hard-q14
+
+## Question Text
+
+As part of a larger virtual reality program, you want to build more life-like objects. Make a Java class called 
+**Bottle**.
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q14/QuestionText.md:6]:
```suggestion
**Bottle** with the requirements below:
```

File: src/oop/special_class_method/hard/q14/QuestionText.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+### Bottle Class:
+
+Your `Bottle` class must have the following items:
+
+- Attributes:
+    - A `boolean` called **isFull**
+    - This is not visible from other classes
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q14/QuestionText.md:14]:
```suggestion
    - A `boolean` called **isFull** that is not visible from other classes.
```

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+
+    @Test
+    public void bottleClassHasCorrectConstructor() throws Throwable {
+        Object classInstance = testClass.createInstance();
+        _assertEquals(false, testClass.getFieldValue(classInstance, attributeName),
+                "Your " + className + " constructor does not initialize the " + attributeName + " attribute to the correct value.");
+    }
+
+    @Test
+    public void getIsFullIsDefinedCorrectly() {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q14/MainTest.java:67]: You can remove this test.

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    public void bottleClassHasRequiredConstructor() {
+        Class<?>[] classArguments = {};
+        assertTrue(testClass.hasConstructor(classArguments),
+                "Your " + className + " constructor does not have the correct parameters.");
+        assertTrue(testClass.hasModifier(classArguments, "public"),
+                "Your " + className + " constructor does not have the correct visibility modifiers.");
+    }
+
+    @Test
+    public void bottleClassHasCorrectConstructor() throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q14/MainTest.java:60]: I would do a repeated test here, just run it 3 or 5 times to make sure it is always false.

File: src/oop/special_class_method/hard/q14/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+        assertTrue(testClass.hasReturnType(getAttributeMethodName, null, boolean.class), String.format(incorrectReturnType, getAttributeMethodName, className));
+        Object testInstance = testClass.createInstance();
+        testClass.setFieldValue(testInstance, value, attributeName);
+        Object getMethodOutput = testClass.callMethod(getAttributeMethodName, testInstance);
+        _assertEquals(value, getMethodOutput,
+                "Your " + getAttributeMethodName + " method does not return the value of the " + attributeName + " attribute.");
+    }
+
+    @Test
+    public void setIsFullIsDefinedCorrectly() {
```
Comment (Hedgemon4) [inline src/oop/special_class_method/hard/q14/MainTest.java:93]: This can also be removed.

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Pull request merged**